Add type annotation into untyped functions automatically

ulwlu opened the pull request:
## Pull Request Check List

- [x] A news fragment is added in `news/` describing what is new.
- [x] Test cases added for changed code.

## Describe what you have changed in this PR

- Add type annotation into untyped functions
- It is because it reduces bug by static type checking such as mypy
- I achieve this by automatically so that I don't missing untyped def
- Related to https://github.com/pdm-project/pdm/issues/261

## Remarks

I believe **adding type annotations into all variables** is **anti-pattern**. 

It is because,
- It occurs many **circular import error** because python is structured only as `parent-to-child`. If you import type from other module, but that other module requires others, then it errors out.
- mypy also doesn't have any kind of options to disallow untyped variable. (Ref: https://stackoverflow.com/questions/58881389/is-there-a-way-for-mypy-to-warn-about-missing-explicit-typing-when-declaring-a-v). Instead of it, mypy recommends to disallow untyped function.

So, I only add typing into functions.

## How to use pyannotate

[pyannotate](https://github.com/dropbox/pyannotate) is made by Guido and his team in dropbox. It's easy and powerful to use compared to [monkeytype](https://github.com/Instagram/MonkeyType) by instagram because monkeytype doesn't support pytest integration officially 

Add below lines into tests/conftest.py. (I don't add this code in this PR, because I think this requires only this time.)
```python
def pytest_collection_finish(session):
    from pyannotate_runtime import collect_types
    collect_types.init_types_collection()

@pytest.fixture(autouse=True)
def collect_types_fixture():
    from pyannotate_runtime import collect_types
    collect_types.start()
    yield
    collect_types.stop()

def pytest_sessionfinish(session, exitstatus):
    from pyannotate_runtime import collect_types
    collect_types.dump_stats("type_info.json")
```

Then run `pytest`. 
It generates the typing informations of all functions what pytest run through into type_info.json.
It is based on sys.setprofile.

When you want to add typing into actual code, run below.

```bash
pdm run pyannotate -w --py3 ./**/*.py
```

The code will have typing automatically.

ulwlu commented on setup.cfg:
```diff
@@ -38,3 +38,9 @@ ignore_errors = true
+disallow_untyped_calls = True
+disallow_untyped_defs = True
+disallow_untyped_decorators = True
+no_implicit_optional = True
```
I use these options with reference to some famous OSS, such as below.

https://github.com/dropbox/mypy-protobuf/blob/master/mypy.ini

frostming replied:
Please add a new line at EOF

ulwlu replied:
fixed here

2ac17a4

ulwlu commented on pyproject.toml:
```diff
@@ -65,6 +67,8 @@ release = "python tasks/release.py"
 doc = {shell = "cd docs && mkdocs serve", help = "Start the dev server for doc preview"}
 lint = "pre-commit run --all-files"
+format = {shell = "black ./pdm ./tests && isort ./pdm ./tests && flake8 ./pdm ./tests"}
+typecheck = "mypy ./pdm ./tests"
```
Actually, it fails with 1 error currently.

![ss 5](https://user-images.githubusercontent.com/41639488/112845275-a6793480-90df-11eb-81f2-83b164f38018.png)

It looks not fatal, but I want to fix in ideal way possibly by another PR.

frostming replied:
Not true, it turns out there are ~500 errors to fix, which is a huge amount

ulwlu replied:
It was actually blocking mypy.
After I excluded that directory -> and removed tests from mypy checking, it shows all errors correctly.

I first of all fixed "missing a type" errors in this PR.
It used to cause "missing a type" errors more than 50 functions, Now all errors of them are gone.

frostming commented on pyproject.toml:
```diff
@@ -65,6 +67,8 @@ release = "python tasks/release.py"
 test = "pytest tests/"
 doc = {shell = "cd docs && mkdocs serve", help = "Start the dev server for doc preview"}
 lint = "pre-commit run --all-files"
+format = {shell = "black ./pdm ./tests && isort ./pdm ./tests && flake8 ./pdm ./tests"}
```
It should be covered by the lint script, I integrated all linting tools into pre-commit, so this shortcut is not needed

ulwlu replied:
sorry, fixed here

da286b7

frostming commented on tests/conftest.py:
```diff
@@ -34,14 +45,20 @@
 
 class LocalFileAdapter(requests.adapters.BaseAdapter):
-    def __init__(self, base_path):
+    def __init__(self, base_path: PosixPath) -> None:
```
```suggestion
    def __init__(self, base_path: Path) -> None:
```

ulwlu replied:
I wondered if Path is correct since LocalFileAdapter is used only by `get_local_finder`, but as thinking of scalability I'm going to fix this.

ulwlu replied:
fixed here

thank you for using suggested change, but it may be differ from flake8 and I wanted if mypy pass through, I fixed on local.

[`271559d` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/271559d31a7363e60d2c95e51972ad9654e18fbb#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R58)

frostming commented on pdm/core.py:
```diff
@@ -99,7 +99,13 @@ def ensure_project(
+        self,
+        args: List[str] = None,
+        prog_name: str = None,
+        obj: Optional[object] = None,
```
```suggestion
        obj: Optional[Project] = None,
```

ulwlu replied:
fixed here

[`271559d` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/271559d31a7363e60d2c95e51972ad9654e18fbb#diff-3a198f24ba7555b1c4cf4d06730a48e6d7494772cda928b78676a76f0f1a0cd9R106)

frostming commented on tests/conftest.py:
```diff
@@ -129,72 +148,79 @@ def _find_candidates(self, requirement: Requirement) -> Iterable[Candidate]:
         self.dependencies = []
 
-    def requires(self, extras=()):
+    def requires(self, extras: Tuple[()] = ()) -> List[NamedRequirement]:
```
```suggestion
    def requires(self, extras: Sequence[str] = ()) -> List[NamedRequirement]:
```

ulwlu replied:
fixed here

[`271559d` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/271559d31a7363e60d2c95e51972ad9654e18fbb#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R188)

frostming commented on tests/test_project.py:
```diff
@@ -81,7 +87,7 @@ def test_project_use_venv(project):
 
 
-def test_project_with_combined_extras(fixture_project):
+def test_project_with_combined_extras(fixture_project: Callable) -> None:
```
```suggestion
def test_project_with_combined_extras(fixture_project: Callable[[str], TestProject]) -> None:
```

ulwlu replied:
fixed here

[`271559d` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/271559d31a7363e60d2c95e51972ad9654e18fbb#diff-6b65416e32913ecea8ccecf110d168f15dede789cc73e8771302190f6da433b1R90-R92)

frostming commented on tests/test_project.py:
```diff
@@ -58,14 +64,14 @@ def test_project_sources_overriding(project):
 
 
-def test_global_project(tmp_path, core):
+def test_global_project(tmp_path: PosixPath, core: Core) -> None:
```
```suggestion
def test_global_project(tmp_path: Path, core: Core) -> None:
```

ulwlu replied:
fixed here

[`271559d` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/271559d31a7363e60d2c95e51972ad9654e18fbb#diff-6b65416e32913ecea8ccecf110d168f15dede789cc73e8771302190f6da433b1R67)

frostming commented on tests/test_plugin.py:
```diff
@@ -17,25 +21,27 @@ def handle(self, project, options) -> None:
 
 
-def make_entry_point(plugin):
+def make_entry_point(plugin: Callable) -> Any:
```
```suggestion
def make_entry_point(plugin: Callable[[Core], Any]) -> Any:
```

ulwlu replied:
fixed here

[`271559d` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/271559d31a7363e60d2c95e51972ad9654e18fbb#diff-b62266dbe51dfb9358f9f7174a4dd8f165eb650d71cf59c42f3a933078e4a95bR36)

frostming reviewed (changes requested):
Just a quick look, there must be many more untyped testing functions. But I don't see it necessary to annotate testing funcs, it doesn't seem conventional.

ulwlu commented:
Oh I'm sorry for having you review many things.
I'm going to check the codes where you pointed out again.

ulwlu commented:
> there must be many more untyped testing functions

Sorry, I noticed that now. 
For example, `test_parse_local_directory_metadata` is not typed, but pyannotate found it for sure.

[type_info.json.log](https://github.com/pdm-project/pdm/files/6225707/type_info.json.log)

I maybe made mistake when applying, so I'm going to dig and add more fix commit in this PR.

frostming commented on pdm/cli/commands/run.py:
```diff
@@ -84,7 +86,9 @@ def _run_command(
-    def _normalize_script(self, script):
+    def _normalize_script(
+        self, script: Any
+    ) -> Tuple[str, Union[Array, String], Union[Dict[str, String], Dict[str, Table]]]:
```
```suggestion
    ) -> Tuple[str, Union[List[str], str], Dict[str, Union[str, dict]]:
```
Avoid using internal types from `tomlkit`

ulwlu replied:
Fixed here. I also removed all internal types from tomlkit from codes.

[`cec10b8` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/cec10b86b17254eebcd050e47ae8bda258b65a27)

frostming commented on pdm/cli/utils.py:
```diff
@@ -125,18 +130,20 @@ def format_help(self):
 
-    def __init__(self, name, version, requirements):
+    def __init__(
+        self, name: str, version: String, requirements: Dict[str, NamedRequirement]
```
```suggestion
        self, name: str, version: String, requirements: Dict[str, Requirement]
```

frostming commented on pdm/formats/legacy.py:
```diff
@@ -11,9 +15,10 @@
 
 
-def check_fingerprint(project, filename):
+def check_fingerprint(project: Union[Project, Project], filename: Path) -> bool:
```
```suggestion
def check_fingerprint(project: Project, filename: os.PathLike) -> bool:
```

ulwlu replied:
Fixed here. I also changed other typings with using filename by same way into os.PathLike.

[`cec10b8` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/cec10b86b17254eebcd050e47ae8bda258b65a27#diff-1fcf2cb1158dfd29bd93f63b19a1923824a632f0daa5209034e320b60fe41e64R22)

frostming commented on pdm/formats/flit.py:
```diff
@@ -12,9 +14,10 @@
 
 
-def check_fingerprint(project, filename):
+def check_fingerprint(project: Optional[Project], filename: Union[Path, str]) -> bool:
```
```suggestion
def check_fingerprint(project: Optional[Project], filename: os.PathLike) -> bool:
```

ulwlu replied:
ditto

frostming commented on pdm/formats/legacy.py:
```diff
@@ -29,48 +34,50 @@ def check_fingerprint(project, filename):
 class LegacyMetaConverter(MetaConverter):
     @convert_from("author")
-    def authors(self, value):
+    def authors(self, value: str) -> Array:
```
```suggestion
    def authors(self, value: str) -> list:
```

ulwlu replied:
ditto (I also removed all internal types from tomlkit from codes.)

frostming commented on pdm/formats/legacy.py:
```diff
@@ -29,48 +34,50 @@ def check_fingerprint(project, filename):
 
     @convert_from("maintainer")
-    def maintainers(self, value):
+    def maintainers(self, value: List[str]) -> Array:
```
```suggestion
    def maintainers(self, value: List[str]) -> list:
```

ulwlu replied:
ditto (I also removed all internal types from tomlkit from codes.)

frostming commented on pdm/formats/legacy.py:
```diff
@@ -29,48 +34,50 @@ def check_fingerprint(project, filename):
 
     @convert_from("license")
-    def license(self, value):
+    def license(self, value: str) -> InlineTable:
```
```suggestion
    def license(self, value: str) -> Dict[str, str]:
```

ulwlu replied:
ditto (I also removed all internal types from tomlkit from codes.)

frostming commented on pdm/formats/legacy.py:
```diff
@@ -29,48 +34,50 @@ def check_fingerprint(project, filename):
 
     @convert_from("source")
-    def source(self, value):
+    def source(self, value: List[Dict[str, Union[bool, str]]]) -> None:
```
```suggestion
    def source(self, value: List[Source]) -> None:
```

```python
from pdm._types import Source
```

frostming commented on pdm/formats/legacy.py:
```diff
@@ -29,48 +34,50 @@ def check_fingerprint(project, filename):
 
     @convert_from("dependencies")
-    def dependencies(self, value):
+    def dependencies(self, value: Dict[str, str]) -> Array:
```
```suggestion
    def dependencies(self, value: Dict[str, str]) -> List[str]:
```

frostming commented on pdm/formats/legacy.py:
```diff
@@ -80,7 +87,7 @@ def dependencies(self, value):
 
     @convert_from("dev-dependencies", name="dev-dependencies")
-    def dev_dependencies(self, value):
+    def dev_dependencies(self, value: Dict) -> Array:
```
```suggestion
    def dev_dependencies(self, value: Dict) -> List[str]:
```

ulwlu replied:
ditto (I also removed all internal types from tomlkit from codes.)

frostming reviewed (commented):
Too much to review, I suggest some of them, others should be similar

frostming commented:
@ulwlu thanks for your effort in adding typing, but can we just leave `tests/` untyped? Most famous projects seem to not annotate `tests/` directory.

This will dramatically reduce the work to review this PR.

ulwlu commented:
@frostming 

> Too much to review

Hi, I am sorry for causing you trouble by adding to many files.

I removed typing from tests, and fixed the all codes with based on what you pointed out.

I believed this would make the project more maintainable, and I found myself hooked into adding typing, but if there are parts you don't want, feel free to point out.

frostming commented on pdm/formats/poetry.py:
```diff
@@ -135,11 +147,20 @@ def dependencies(self, source):
+            str,
+            Union[
+                Union[Dict[str, str], List[str], str],
+                Union[Dict[str, str], List[str], str],
```
Need to deduplicate here

frostming commented on pdm/formats/requirements.py:
```diff
@@ -1,17 +1,28 @@
 
-def _requirement_to_str_lowercase_name(requirement):
+def _requirement_to_str_lowercase_name(
+    requirement: Union[VcsRequirement, Requirement]
```
```suggestion
    requirement: Requirement
```

frostming commented on pdm/formats/requirements.py:
```diff
@@ -1,17 +1,28 @@
 
 from distlib.wheel import Wheel
+from pip._internal.index.package_finder import PackageFinder
+from pip._internal.req.req_install import InstallRequirement
```
Import from `pdm.models.pip_shims` please

ulwlu replied:
fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-508798c1e0391fc888574533cc635c84f3be7f0a6732a439eafdc5bd2581da41R15)

and I checked all `pip._internal` importing not from `pdm.models.pip_shims`.

Some codes are still using `pip._vendor` such as `from pip._vendor.requests.models import Response`, and if you'd like to move all of them into pdm.models, pleate tell me.

frostming commented on pdm/installers/_editable_install.py:
```diff
@@ -2,9 +2,10 @@
 
 
-def install(setup_py, prefix, lib_dir, bin_dir):
+def install(setup_py: Path, prefix: Path, lib_dir: Path, bin_dir: Path) -> None:
```
```suggestion
def install(setup_py: str, prefix: str, lib_dir: str, bin_dir: str) -> None:
```
These args are taken from `sys.argv`, apparently they are strings

frostming commented on pdm/installers/synchronizers.py:
```diff
@@ -18,23 +21,23 @@
         return self._result
 
-    def exception(self):
+    def exception(self) -> Optional[Any]:
```
```suggestion
    def exception(self) -> Optional[Exception]:
```

ulwlu replied:
Fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-3c7dbb09445dee162dcefb5ac080ace1d0becda785d2c3fb9cab128e5d94632cR37)

frostming commented on pdm/installers/synchronizers.py:
```diff
@@ -78,7 +81,11 @@ def __init__(
+    def create_executor(
+        self,
+    ) -> Iterator[
+        Union[Iterator, Iterator[ThreadPoolExecutor], Iterator[DummyExecutor]]
```
Doesn't make sense, should be
`Union[Iterator[ThreadPoolExecutor], Iterator[DummyExecutor]]`

ulwlu replied:
Fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-3c7dbb09445dee162dcefb5ac080ace1d0becda785d2c3fb9cab128e5d94632cR86)

frostming commented on pdm/installers/synchronizers.py:
```diff
@@ -18,23 +21,23 @@
         self._exc = None
 
-    def set_result(self, result):
+    def set_result(self, result: Candidate) -> None:
```
```suggestion
    def set_result(self, result: Any) -> None:
```

ulwlu replied:
Fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-3c7dbb09445dee162dcefb5ac080ace1d0becda785d2c3fb9cab128e5d94632cR28)

frostming commented on pdm/models/environment.py:
```diff
@@ -141,13 +141,13 @@ def activate(self):
             pkg_resources.working_set = _old_ws
 
-    def is_local(self, path) -> bool:
+    def is_local(self, path: Path) -> bool:
```
```suggestion
    def is_local(self, path: os.PathLike) -> bool:
```

ulwlu replied:
fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-b36c5d0a379286c6493f3b10d4e82a66fe1e48011170ac20c6990045d4f0a3b5R144)

frostming commented on pdm/models/environment.py:
```diff
@@ -367,7 +367,7 @@ def get_paths(self) -> Dict[str, str]:
         return paths
 
-    def is_local(self, path) -> bool:
+    def is_local(self, path: Path) -> bool:
```
```suggestion
    def is_local(self, path: os.PathLike) -> bool:
```

ulwlu replied:
fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-b36c5d0a379286c6493f3b10d4e82a66fe1e48011170ac20c6990045d4f0a3b5R370)

frostming commented on pdm/models/markers.py:
```diff
@@ -120,7 +121,13 @@ def split_marker_extras(
+    markers: Union[
+        List[Tuple[Variable, Op, Value]],
+        List[Union[Tuple[Variable, Op, Value], str]],
+        Tuple[Variable, Op, Value],
```
For such type annotations which is too verbose, I would like to remove them.

ulwlu replied:
I get it.
I'll look the codes again and reduce too verbose annotations.

ulwlu replied:
fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-79c8eb95336c3a619cac203017451bbd32680ac013513a013b2c38087fc6ed87R123)

I checked how markers should be typed here.
https://gitlab.u-angers.fr/projetsetudiants/3a_2021_trafic2/-/blob/f593f8e345fec6a206816b8212675af186e06c9f/Traffik0.1/Lib/site-packages/pip/_vendor/packaging/markers.py#L299

frostming commented on pdm/models/requirements.py:
```diff
@@ -111,7 +111,7 @@ def marker(self) -> Optional[Marker]:
 
     @marker.setter
-    def marker(self, value) -> None:
+    def marker(self, value: Union[Marker, None]) -> None:
```
```suggestion
    def marker(self, value: Optional[Marker]) -> None:
```

ulwlu replied:
Fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-86e5c34f0c74cee12fced7a20255f336f7d232d1363cc1f3ac978d2d1788f960R114)

frostming commented on pdm/models/specifiers.py:
```diff
@@ -203,13 +205,13 @@ def _rearrange(
         return hash(self._comp_key())
 
-    def __eq__(self, other):
+    def __eq__(self, other: PySpecSet) -> bool:
```
```suggestion
    def __eq__(self, other: Any) -> bool:
```

ulwlu replied:
I thought this one is supposed to have PySpecSet from seeing the code, but I now realized __eq__ will be used for comparison of anything. I'll fix.

```python
    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, PySpecSet):
            return False
        return self._comp_key() == other._comp_key()
```

ulwlu replied:
Fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-8b1134a03683f14d179d12de9b62c6e35d2689782669129428ad0324c77ff0abR214)

frostming commented on pdm/pep582/sitecustomize.py:
```diff
@@ -1,10 +1,12 @@
-def get_pypackages_path(maxdepth=5):
-    def find_pypackage(path, version):
+def get_pypackages_path(maxdepth: int = 5) -> Path:
+    def find_pypackage(path: Path, version: str) -> Union[None, Path]:
```
Don't type this file, it can be run by Python 2.7

ulwlu replied:
Fixed here. Sorry I didn't notice it.

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-e5d05cfeca8df090e45a786f1e58cac1d4fc31c17897dcb8d37bae44ababf47fR6-R7)

frostming commented on pdm/project/config.py:
```diff
@@ -21,7 +21,9 @@ def load_config(file_path: Path) -> Dict[str, Any]:
 
-    def get_item(sub_data):
+    def get_item(
+        sub_data: Union[Dict[str, bool], Dict[str, Dict], Dict]
```
```suggestion
        sub_data: Dict[str, Any]
```

ulwlu replied:
Fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-201cb9c73ada491d4dc9fd0bdb499a2cebf4ff29882c693f3a99490c145d43b8R24)

frostming commented on pdm/project/config.py:
```diff
@@ -21,7 +21,9 @@ def load_config(file_path: Path) -> Dict[str, Any]:
-    def get_item(sub_data):
+    def get_item(
+        sub_data: Union[Dict[str, bool], Dict[str, Dict], Dict]
+    ) -> Union[Dict[str, bool], Dict[str, str]]:
```
```suggestion
    ) -> Dict[str, Any]:
```

ulwlu replied:
Fixed here

[`312287c` (#354)](https://github.com/pdm-project/pdm/pull/354/commits/312287c375db739caa180500afbc976002515113#diff-201cb9c73ada491d4dc9fd0bdb499a2cebf4ff29882c693f3a99490c145d43b8R24)

frostming commented on pdm/formats/base.py:
```diff
@@ -29,7 +29,9 @@ class Unset(Exception):
 class _MetaConverterMeta(type):
-    def __init__(cls, name, bases, ns):
+    def __init__(
+        cls, name: str, bases: Dict[str, Any], ns: Union[None, Path, str]
```
This annotation seems wrong.

ulwlu commented:
@frostming 
Hi, thank you for reviewing.
I fixed where you pointed out and other codes where have similar problem.

I also wrote one question.

> Some codes are still using pip._vendor such as from pip._vendor.requests.models import Response, and if you'd like to move all of them into pdm.models, pleate tell me.

frostming commented on pdm/formats/requirements.py:
```diff
@@ -116,7 +127,11 @@ def convert(project, filename, options):
-def export(project, candidates, options):
+def export(
+    project: Project,
+    candidates: Union[List[Candidate], List[NamedRequirement]],
```
```suggestion
    candidates: Union[List[Candidate], List[Requirement]],
```

frostming commented:
That is really huge, if you don't mind I will merge it to a temporary branch and do more modifications based on that.

ulwlu commented:
Sure, thank you very much.

Since I see you start applying suggestions, I stop modifying here.
